client/ayon_core/pipeline/workfile/workfile_template_builder.py

```diff
@@ -1529,35 +1529,22 @@ def _before_repre_load(self, placeholder, representation):
 
         pass
 
-    def _reduce_last_version_repre_entities(self, representations):
-        """Reduce representations to last verison."""
-
-        mapping = {}
-        # TODO use representation context with entities
-        # - using 'folder', 'subset' and 'version' from context on
-        #   representation is danger
-        for repre_entity in representations:
-            repre_context = repre_entity["context"]
-
-            folder_name = repre_context["asset"]
-            product_name = repre_context["subset"]
-            version = repre_context.get("version", -1)
-
-            if folder_name not in mapping:
-                mapping[folder_name] = {}
-
-            product_mapping = mapping[folder_name]
-            if product_name not in product_mapping:
-                product_mapping[product_name] = collections.defaultdict(list)
-
-            version_mapping = product_mapping[product_name]
-            version_mapping[version].append(repre_entity)
+    def _reduce_last_version_repre_entities(self, repre_contexts):
+        """Reduce representations to last version."""
+
+        version_mapping_by_product_id = {}
+        for repre_context in repre_contexts:
+            product_id = repre_context["product"]["id"]
+            version = repre_context["version"]["version"]
+            version_mapping = version_mapping_by_product_id.setdefault(
+                product_id, {}
+            )
+            version_mapping[version].append(repre_context)
```

Comment:
This logic here fails - see [comment](https://github.com/ynput/ayon-core/pull/416/files#r1569624798)
Not sure if this should be by `version` or `product_id` - but I suppose by product id since we're trying to reduce to the version?

Either this is something like:
```suggestion
            version_mapping = version_mapping_by_product_id.setdefault(
                product_id, {}
            )
            version_mapping.setdefault(version, []).append(repre_context)
```
Or:

```suggestion
            version_mapping = version_mapping_by_product_id.setdefault(
                product_id, []
            )
            version_mapping.append(repre_context)
```
However, I'm not entirely sure of the intent of the reduce function.

Reply:
This may be simpler if:
```python
version_mapping_by_product_id = defaultdict(lambda: defaultdict(list))
```

Then the logic here would just be:
```python
version_mapping_by_product_id[product_id][version].append(repre_context)
```